parsing/builtin_attributes.ml

```diff
@@ -101,6 +101,29 @@ let rec deprecated_of_str = function
   | _ -> None
 
 
+let rec check_nested_match e =
+  if List.for_all
+    (function
+      | ({txt="ocaml.delimited"; _}, _) -> false
+      | _ -> true
+    )
+    e.pexp_attributes
+  then match e.pexp_desc with
+    | Pexp_let (_, _, e)
+    | Pexp_letmodule (_, _, e)
+    | Pexp_letexception (_, e)
+    | Pexp_ifthenelse (_, e, None)
+      -> check_nested_match e
+    | Pexp_ifthenelse (_, e1, Some e2)
+    | Pexp_sequence (e1, e2) ->
+        check_nested_match e1; check_nested_match e2
+
+    | Pexp_match _
+    | Pexp_function _
+    | Pexp_try _ ->
+        Location.prerr_warning e.pexp_loc Warnings.Pattern_matching_should_be_delimited
+    | _ -> ()
```

Comment:
I'd make the match explicitly exhaustive (i.e. listing all cases instead of using `_`) so when people add new constructors to `expression_desc` they think about whether they need to check some subtree or not.